File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left = 1
        right = len(numbers)
        for i in range(len(numbers)-1):
            # case 1: left + right > target
            if numbers[left-1] + numbers[right-1] > target:
                right -= 1
            # case 2: left + right < target
            elif numbers[left-1] + numbers[right-1] < target:
                left += 1
            else:
                break
        return [left, right]

File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // first loop
        int left = 0, right = matrix.size()-1;
        while(left <= right){
            int kiri = 0, kanan = matrix[0].size()-1;
            int mid = (left + right) / 2;
            if(target > matrix[mid][kanan]){
                left = mid+1;
            }
            else if(target < matrix[mid][kiri]){
                right = mid-1;
            }
            else{
                while(kiri <= kanan){
                    int tengah = (kiri+kanan) / 2;
                    if(matrix[mid][tengah] == target){
                        return true;
                    }
                    else if(matrix[mid][tengah] < target){
                        kiri = tengah+1;
                    }
                    else{
                        kanan = tengah-1;
                    }
                }
                return false;
            }
        }
        return false;
    }
};
